Mathlib/LinearAlgebra/Matrix/FixedDetMatrices.lean

```diff
@@ -0,0 +1,56 @@
+/-
+Copyright (c) 2024 Chris Birkbeck. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Birkbeck
+-/
+
+import Mathlib.LinearAlgebra.Matrix.SpecialLinearGroup
+
+/-!
+# Matrices with fixed determinant
+
+This file defines the type of matrices with fixed determinant `m` and proves some basic results
+about them.
+
+Note: Some of this was done originally in Lean 3 in the kbb repository, so credit to those authors.
+-/
+
+variable (n : Type*) [DecidableEq n] [Fintype n] (R : Type*) [CommRing R]
+
+/--The set of matrices with fixed determinant `m`. -/
+def FixedDetMatrices (m : R) := { A : Matrix n n R // A.det = m }
+
+namespace FixedDetMatrices
+
+open ModularGroup Matrix SpecialLinearGroup MatrixGroups
+
+lemma ext' (m : R) {A B : FixedDetMatrices n R m} (h : A.1 = B.1) : A = B := by
+  cases A; cases B
+  congr
+
+@[ext]
+lemma ext (m : R) {A B : FixedDetMatrices n R m} (h : ∀ i j , A.1 i j = B.1 i j) : A = B := by
+  apply ext'
+  ext i j
+  apply h
+
+instance (m : R) : HSMul (SpecialLinearGroup n R) (FixedDetMatrices n R m)
```

Comment:
Can this not be a plain `SMul`?

Reply:
when I do this I get the following error which I dont know how to fix: 

```
function expected at
  SMul (SpecialLinearGroup n R) (FixedDetMatrices n R m)
term has type
  Type (max u_2 u_1)
```